File: javascript/051.js
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function(n) {
  const n2 = n * n;
  const board = (new Array(n2)).fill(false);

  const isPosValid = (row, col) => {
    // check all rows with col
    for (let i = 0; i < n; ++i) {
      if (board[i * n + col] && i !== row) return false;
    }

    // check all cols with row
    for (let i = 0; i < n; ++i) {
      if (board[row * n + i] && i !== col) return false;
    }

    // check 45 direction
    var row1 = row >= col ? row - col : 0;
    var col1 = row >= col ? 0 : col - row;
    while(row1 < n && col1 < n) {
      if (row1 !== row && board[row1 * n + col1]) return false;
      ++row1;
      ++col1;
    }

    // check 135 direction
    var row2 = row + col >= n - 1 ? row + col - n + 1 : 0;
    var col2 = row + col >= n - 1 ? n - 1 : row + col;
    while(row2 < n && col2 >= 0) {
      if (row2 !== row && board[row2 * n + col2]) return false;
      ++row2;
      --col2;
    }

    return true;
  };

  let solutions = [];
  findSolutions = row => {
    for(let col = 0; col < n; ++col) {
      if (!isPosValid(row, col)) continue;

      board[row * n + col] = true;
      if (row < n - 1) {
        findSolutions(row + 1);
      } else {
        solutions.push(Array.from(board));
      }
      board[row * n + col] = false;
    }
  }

  findSolutions(0);

  toResult = solutions => {
    let strArrayArray = [];
    solutions.forEach(solution => {
      let strArray = [];
      for(let i = 0; i < n; ++i) {
        let str = "";
        for (let j = 0; j < n; ++j) {
          str += (solution.shift() ? "Q" : ".");
        }
        strArray.push(str);
      }
      strArrayArray.push(strArray);
    });

    return strArrayArray;
  }
  return toResult(solutions);
};

File: ruby/051.rb
# @param {Integer} n
# @return {String[][]}
def is_valid(board, row, col)
  return false if board[row].each_with_index.any? do |item, c|
    c != col && item == "Q"
  end

  return false if board.each_with_index.any? do |line, r|
    r != row && line[col] == "Q"
  end

  if row >= col
    return false if (row - col).upto(board.length - 1).any? do |r|
      c = r - row + col
      r != row && board[r][c] == "Q"
    end
  else
    return false if 0.upto(board.length - 1 - col + row).any? do |r|
      c = col - row + r
      r != row && board[r][c] == "Q"
    end
  end

  if row + col <= board.length
    return false if 0.upto(row + col).any? do |r|
      c = row + col - r
      r != row && board[r][c] == "Q"
    end
  else
    return false if (row + col - board.length + 1).upto(board.length - 1).any? do |r|
      c = row + col - r
      r != row && board[r][c] == "Q"
    end
  end

  true
end

def find_solution board, row, col
  if row == board.length
    sum = 0
    board.each do |line|
      line.each do |item|
        sum += 1 if item == "Q"
      end
    end
    return sum == board.length
  end

  return find_solution(board, row + 1, 0) if col == board.length
  return find_solution(board, row, col + 1) unless is_valid(board, row, col)
  board[row][col] = "Q"
  if find_solution(board, row, col + 1)
    return true
  else
    board[row][col] = "*"
    return find_solution(board, row, col + 1)
  end
end

def solve_n_queens(n)
  board = Array.new(n, Array.new(n, "*"))
  if find_solution(board, 0, 0)
    return board
  else
    return nil
  end
end

p solve_n_queens(4)